keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+Application.
+
+#### Retrieving Applications
+In order to retrieve the Application corresponding to an API object, a client should perform the following.
+
+1. Examine the object for a `kubernetes.io/application` annotation.
+1. If no annotation is found, the object is not associated with an Application.
+1. If the annotation is found, retrieve the Application object with the indicated name from the same namespace of the 
+object.
+1. If no such Application is found, the Application is either deleted or has not yet been created.
+1. If the Application is found, validate that the `Spec.Selector` of the Application object matches the labels of the 
+object. If the object's labels do not match, the object has been orphaned.
+1. Check the `OwnerReferences` of the object for a reference to the Application by `UID` to ensure that ownership has 
+been established.
+
+#### Retrieving Dependencies
+In order to retrieve the `Spec.Dependencies` of an Application, a client should perform the following.
+
+1. For each name in the `Spec.Dependencies` list, retrieve the Application of that name, in the namespace of the 
+Application under consideration. If no such Application is found, the dependency is not installed.
+
+#### Health Checks
+The semantics of readiness for an Application vary too greatly for declarative specification. We propose to encode 
+the readiness semantics of Applications in the same manner as Pods (i.e. by using programs). An ApplicationHealthCheck 
+object contains the template for a Pod that encodes a semantic readiness check for the components of an Application. 
+The requirements of implementing a ApplicationHealthCheck Pod are as follows.
+
+1. A health check Pod must implement a readiness check.
+1. The readiness check of the Pod may only return successfully when the conditions of readiness are met for the entire 
+application. 
+1. An ApplicationHealthCheck Pod should not supply other application telemetry. When its readiness check fails that 
+telemetry would be unavailable (at precisely the time it is most necessary).
+
+### Application Controller
+The Application Controller will be responsible for configuring ownership to enable GC (Garbage Collection) for an 
+Application's components and for updating the `Status` of the Application with respect to installation 
+status of the components and dependencies of the Application and to the readiness of the Application's health check.
+
+#### Garbage Collection
+In order to facilitate GC for Application components, an  `OwnerRefernce` from the API object must be 
+added to all of its indicated component objects. As an Application can be composed of arbitrary Kinds, like the 
+Garbage Collector, the Application Controller will have to use the discovery API to determine when new Kinds are 
+registered with the API server. It will have to watch all Kinds that may be included in an Application's `Components`.
+
```

Comment:
There are some "components" that we in the Helm community have received lots of feedback on, when it comes to deleting. PVCs are a good example: Many operators do NOT want their PVCs deleted when an application is deleted. (Of course, others do...) You may need to facilitate some way of annotating these as "do not delete".

Reply:
FWIW I never really understood the issue about PVCs. What you really want to keep is the PVs, and so the cluster administrator should change the `reclaimPolicy` on StorageClasses to `Retain`.